app/Http/Controllers/ProfileController.php

```diff
@@ -4,40 +4,76 @@
 
 use App\Http\Requests\ProfileRequest;
 use App\Http\Requests\ProfileUploadRequest;
+use App\Repositories\ProfileRepository;
 use Illuminate\Http\JsonResponse;
 use Illuminate\Http\Request;
-use App\Repositories\ProfileRepository;
 
 class ProfileController extends Controller
 {
     protected $repo;
 
-    public function __construct(ProfileRepository $repo) {
+    public function __construct(ProfileRepository $repo)
+    {
         $this->repo = $repo;
     }
 
-    public function getProfile(Request $request, $address) : JsonResponse {
-        return response()->json($this->repo->getProfile($request,$address));
+    public function getProfile(Request $request, $address): JsonResponse
+    {
+        $profile = $request->user();
+        $addressProfile = $this->repo->getProfile($address, ['users.circle.protocol', 'users.teammates', 'users.histories.epoch']);
+        if ($profile && !$profile->admin_view) {
+            if (!$addressProfile || count(array_intersect($profile->circle_ids(), $addressProfile->circle_ids())) < 1) {
+                return response()->json(['message' => 'User has no permission to view this profile'], 403);
+            }
+        }
+        return response()->json($addressProfile);
     }
 
-    public function updateMyProfile(ProfileRequest $request) {
+    public function updateMyProfile(ProfileRequest $request): JsonResponse
+    {
         return response()->json($this->repo->saveProfile($request));
     }
 
-    public function uploadProfileAvatar(ProfileUploadRequest $request, $address) : JsonResponse {
-        $profile = $this->repo->uploadProfileAvatar($request, $request->address);
-        if($profile)
+    public function uploadMyProfileAvatar(ProfileUploadRequest $request, $address = null): JsonResponse
+    {
+        $profile = $this->repo->uploadProfileAvatar($request);
+        if ($profile)
             return response()->json($profile);
 
-        return response()->json(['message' => 'File Upload Failed' ,422]);
+        return response()->json(['message' => 'File Upload Failed', 422]);
     }
 
-    public function uploadProfileBackground(ProfileUploadRequest $request, $address) : JsonResponse {
-       $profile = $this->repo->uploadProfileBackground($request, $request->address);
-       if($profile)
-           return response()->json($profile);
+    public function uploadMyProfileBackground(ProfileUploadRequest $request, $address = null): JsonResponse
+    {
+        $profile = $this->repo->uploadProfileBackground($request);
+        if ($profile)
+            return response()->json($profile);
 
-        return response()->json(['message' => 'File Upload Failed' ,422]);
+        return response()->json(['message' => 'File Upload Failed', 422]);
     }
 
+    public function manifest(Request $request): JsonResponse
```

Comment:
Instead of how manifest is currently returning users, either remove that, or have it be the current user's users, with profile filled in (if that is easy (I know it is redundant), but ideally the front end wouldn't be juggling so many schemas.

Reply:
Also let's have manifest accept an optional `circle_id` parameter, and return the same things that `fullCircleData` does using either the parameter or the first circle if it exists.